### PR TITLE
INT-3600: FileSplitter Namespace and Docs

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3600

Add `<int-file:splitter/>` namespace component.
Add `FileSplitter` documentation.

### CODE REVIEW

File: spring-integration-file/src/main/resources/org/springframework/integration/file/config/spring-integration-file-4.2.xsd [hunk 1]
```diff
@@ -547,6 +547,87 @@ Only files matching this regular expression will be picked up by this adapter.
...
+				<xsd:simpleType>
+					<xsd:union memberTypes="xsd:boolean xsd:string"/>
+				</xsd:simpleType>
+			</xsd:attribute>
+			<xsd:attribute name="iterator" use="optional" default="true">
+				<xsd:annotation>
+					<xsd:documentation>
+						Set this flag to determine whether the lines are emitted via an iterator (line at a time)
+						or the whole file is read into memory and then the lines emitted. When true, if
+						'apply-sequence' is 'true', the `sequenceSize` header is set to '0'. Default: 'true'.
```
Comment (artembilan) [inline spring-integration-file/src/main/resources/org/springframework/integration/file/config/spring-integration-file-4.2.xsd]: `When 'iterator' and 'apply-sequence' are 'true', the `sequenceSize` header is set to '0'` ???

Minor, of course, and we can live with that. Just want to be sure that I see anything here correctly :smile:

File: spring-integration-file/src/main/resources/org/springframework/integration/file/config/spring-integration-file-4.2.xsd [hunk 2]
```diff
@@ -547,6 +547,87 @@ Only files matching this regular expression will be picked up by this adapter.
...
+					</xsd:documentation>
+				</xsd:annotation>
+				<xsd:simpleType>
+					<xsd:union memberTypes="xsd:boolean xsd:string"/>
+				</xsd:simpleType>
+			</xsd:attribute>
+			<xsd:attribute name="markers" use="optional" default="false">
+				<xsd:annotation>
+					<xsd:documentation>
+						Set this flag to determine whether start of file and end of file markers are emitted
```
Comment (artembilan) [inline spring-integration-file/src/main/resources/org/springframework/integration/file/config/spring-integration-file-4.2.xsd]: ... whether start ~~of file~~ and end of file ...

No need of tautology.

File: spring-integration-file/src/main/resources/org/springframework/integration/file/config/spring-integration-file-4.2.xsd [hunk 3]
```diff
@@ -547,6 +547,87 @@ Only files matching this regular expression will be picked up by this adapter.
...
+						before and after the file content messages respectively. The marker message payloads are
+						of type 'FileSplitter.FileMarker'. When `apply-sequence` is true, the sequence information
```
Comment (artembilan) [inline spring-integration-file/src/main/resources/org/springframework/integration/file/config/spring-integration-file-4.2.xsd]: `When 'apply-sequence' is 'true'`
Wrong quotes and mark "true" as a code for consistency with other similar here.
Of course, minor again.
And I can fix them all on merge.
Just let me know that I look to the correct way.

File: spring-integration-file/src/test/java/org/springframework/integration/file/config/FileSplitterParserTests.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+	@Test
+	public void testComplete() {
+		assertFalse(TestUtils.getPropertyValue(this.splitter, "iterator", Boolean.class));
+		assertTrue(TestUtils.getPropertyValue(this.splitter, "markers", Boolean.class));
+		assertTrue(TestUtils.getPropertyValue(this.splitter, "requiresReply", Boolean.class));
+		assertTrue(TestUtils.getPropertyValue(this.splitter, "applySequence", Boolean.class));
+		assertEquals(Charset.forName("UTF-8"), TestUtils.getPropertyValue(this.splitter, "charset"));
+		assertEquals(5L, TestUtils.getPropertyValue(this.splitter, "messagingTemplate.sendTimeout"));
+		assertEquals(out, TestUtils.getPropertyValue(this.splitter, "outputChannel"));
```
Comment (artembilan) [inline spring-integration-file/src/test/java/org/springframework/integration/file/config/FileSplitterParserTests.java]: Nit-pick: `this.out`

File: src/reference/asciidoc/file.adoc [hunk 5]
```diff
@@ -465,3 +465,83 @@ To configure File specific transformers you can use the appropriate elements fro
...
+By default, the splitter uses an `Iterator` to emit lines one-at-a-time as they are read from the file.
+Setting the `iterator` property to `false` causes it to read all the lines into memory before emitting them as messages.
+One use case for this might be if you want to detect I/O errors on the file before sending any messages containing
+lines.
+However, it is only practical for relatively short files.
+
+Inbound payloads can be `File`, `String` (a `File` path), `InputStream`, or `Reader`.
+Other payload types will be emitted unchanged.
+
+[source, java]
```
Comment (artembilan) [inline src/reference/asciidoc/file.adoc]: Isn't nit-pick and not minor: the code below is XML, not Java :smile:

File: src/reference/asciidoc/file.adoc [hunk 6]
```diff
@@ -465,3 +465,83 @@ To configure File specific transformers you can use the appropriate elements fro
...
+    auto-startup="" <10>
+    order="" <11>
+    phase="" /> <12>
+----
+
+<1> The bean name of the splitter.
+
+<2> Set to `true` to use an iterator (default); `false` to load the file into memory before sending lines.
+
+<3> Set to `true` to emit start of file/end of file marker messages before and after the file data.
```
Comment (artembilan) [inline src/reference/asciidoc/file.adoc]: Yeah. This sound better, I can apply it for XSD as well.

**Comment by artembilan** [comment]
That's all. Wait for the feedback

File: spring-integration-file/src/main/resources/org/springframework/integration/file/config/spring-integration-file-4.2.xsd (hunk 1 above)
Comment (garyrussell) [inline spring-integration-file/src/main/resources/org/springframework/integration/file/config/spring-integration-file-4.2.xsd]: Not sure I understand the complaint here - is this better...

>When this is 'true', if 'apply-sequence' is also 'true', the 'sequenceSize' header is set to '0'. Default: 'true'.

**Comment by garyrussell** [comment]
Rebased, pushed.

**Comment by artembilan** [comment]
Merging...

**Comment by artembilan** [comment]
Merged

**Pull request closed without merging**